### PR TITLE
GH-2605: (S)FTP test cached sessions

### PR DESCRIPTION
**Pull request by garyrussell**
Resolves https://github.com/spring-projects/spring-integration/issues/2605

### CODE REVIEW

File: spring-integration-file/src/test/java/org/springframework/integration/file/remote/session/CachingSessionFactoryTests.java [hunk 1]
```diff
@@ -61,6 +62,7 @@ public void testCacheAndReset() {
 		assertTrue(sess1.isOpen());
 		sess1 = cache.getSession();
 		assertEquals("session:1", TestUtils.getPropertyValue(sess1, "targetSession.id"));
+		assertTrue((TestUtils.getPropertyValue(sess1, "targetSession.testCalled", Boolean.class)));
```
Comment (artembilan) [inline spring-integration-file/src/test/java/org/springframework/integration/file/remote/session/CachingSessionFactoryTests.java]: The `TestSession` is a part of this class, so its `private` properties are visible.
And afterwards we don't need that `@SuppressWarnings("unused")` as well.

File: spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/SftpSession.java [hunk 2]
```diff
@@ -278,4 +278,20 @@ public ChannelSftp getClientInstance() {
...
 	}
 
+	@Override
+	public boolean test() {
+		return isOpen() && doTest();
+	}
+
+	private boolean doTest() {
+		try {
+			this.channel.stat(this.channel.getHome());
```
Comment (artembilan) [inline spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/SftpSession.java]: Why `jschSession.sendKeepAliveMsg()` doesn't work for us?

File: src/reference/asciidoc/ftp.adoc [hunk 3]
```diff
@@ -1412,6 +1413,9 @@ Starting with Spring Integration 3.0, the `CachingConnectionFactory` provides a
 When invoked, all idle sessions are immediately closed and in-use sessions are closed when they are returned to the cache.
 New requests for sessions establish new sessions as necessary.
 
+Starting with version 5.1, the `CachingSessionFactory` has a new poperty `testSession`.
```
Comment (artembilan) [inline src/reference/asciidoc/ftp.adoc]: Typo: p**r**operty

**Review by artembilan — CHANGES_REQUESTED** [review]
Just a couple questions.

File: spring-integration-file/src/test/java/org/springframework/integration/file/remote/session/CachingSessionFactoryTests.java (hunk 1 above)
Comment (garyrussell) [reply spring-integration-file/src/test/java/org/springframework/integration/file/remote/session/CachingSessionFactoryTests.java]: ? 

It's wrapped in a `CachedSession` the `CachedSession.targetSession` is private.

File: spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/SftpSession.java (hunk 2 above)
Comment (garyrussell) [reply spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/SftpSession.java]: That command doesn't wait for a reply so we get false positives (e.g. when pulling the network cable).

**Pull request merged**